lib/lhm/connection.rb

```diff
@@ -0,0 +1,57 @@
+
+module Lhm
+  class Connection
+
+    def initialize(connection:, default_log_prefix: nil, retry_options: {})
+      @default_log_prefix = default_log_prefix
+      @connection = connection
+      @retry_options = retry_options || default_retry_config
+      @sql_retry = Lhm::SqlRetry.new(
+        connection,
+        retry_options,
+      )
+    end
+
+    def execute(query, retry_options = {})
+      exec_with_retries(:execute, query, retry_options)
+    end
+
+    def update(query, retry_options = {})
+      exec_with_retries(:update, query, retry_options)
+    end
+
+    def select_value(query, retry_options = {})
+      exec_with_retries(:select_value, query, retry_options)
+    end
+
+    # Delegate ALL unknown method calls to ActiveRecord's connection.
+    # This ensures that there will be no breaking changes.
+    def method_missing(m, *args, &block)
+      @connection.public_send(m, *args, &block)
+    end
```

Comment:
This would break in Ruby 3 if the underlying method has keyword arguments. There it would need to be either

```ruby
def method_missing(m, *args, **kwargs, &block)
  @connection.public_send(m, *args, **kwargs, &block)
end
```

or with the new argument forwarding operator 

```ruby
def method_missing(m, ...)
  @connection.public_send(m, ...)
end
```

However, the version with `**kwargs` is only supported since Ruby 2.7 and the one with `...` is supported since Ruby 3.0.

There are some ways around this, like defining the method differently based on the Ruby version

```ruby
if RUBY_VERSION < "3"
  def method_missing(m, *args, &block)
    @connection.public_send(m, *args, &block)
  end
else
  def method_missing(m, ...)
    @connection.public_send(m, ...)
  end
end
```

Alternatively, you could also use Ruby's [SimpleDelegator](https://docs.ruby-lang.org/en/3.0.0/SimpleDelegator.html) here. The class would need to subclass SimpleDelegator, you'd need to add a `super(connection)` to the end of your `initialize` and then you should be able to remove the `method_missing` definition. SimpleDelegator is doing more or less the same, but handles the different ruby versions for you.

Reply:
That's a GREAT catch! I'll investigate using a `SimpleDelegator` and see if it fits adequately in the codebase! If it works, I'll add appropriate comments to make sure it's not too confusing for the next contributor.

Reply:
Do we need to use SimpleDelegator to only get around 1 if condition? I think 1 if condition is OK.